### ISSUE TITLE
Suggested Name should allow name starting with a prefix

### ISSUE THREAD

**Issue by jcouv** [dotnet/roslyn#22901]
There are multiple naming conventions which involve prefixes. For instance, `_` for private fields.
But the "suggested name" recommendation doesn't handle such prefixes. In the example below, it should offer `_messageProvider` (amongst others).

![image](https://user-images.githubusercontent.com/12466233/32139611-f5844372-bc02-11e7-94fc-8d0b60225dd8.png)

FYI @CyrusNajmabadi

**Comment by CyrusNajmabadi** [comment]
Hunh.  I thought i already opened a bug on this :)   I think i talked about this with either @dpoeschl or @rchande .  Shouldn't be hard to do.  And we shoudl make it general.  i.e. if i do this:

```Customer _new```

then we should offer ```_newCustomer``` in the list.

**Comment by Pilchie** [comment]
I think it does *if* you have a naming rule configured, but it doesn't sound like you do.

**Comment by rchande** [comment]
Right, this should work if you have an appropriate naming rule configured.

**Comment by DustinCampbell** [comment]
Should it be a bit smarter and recognize that I'm intentionally typing with a prefix?

**Comment by rchande** [comment]
We did talk about that at some point; not sure if there's a bug tracking that request.

**Comment by DustinCampbell** [comment]
*This* could be that very issue. :smile:

**Comment by CyrusNajmabadi** [comment]
Right.  I want this to work *regardless* of the naming rules i've setup.  I may not like _'s.  but if i use one in one of my types, i'd still like good suggestions :)

**Comment by DustinCampbell** [comment]
That's my feeling too. It also might be that the user simply hasn't set their naming rules yet.

**Comment by Pilchie** [comment]
Agreed.

**Comment by sharwell** [comment]
In other words, after each typed character, the naming suggestions should be updated to have the general form:

[prefix][overlapping][suffix]

Where the text typed by the user is [prefix][overlapping], the original suggestion after typing <kbd>Space</kbd> was [overlapping][suffix], and [overlapping] is the longest viable sequence. If [overlapping] is empty and [prefix] is not empty, a word separator should be placed where [overlapping] appears in the example and/or the first character of [suffix] should be case-corrected to meet the naming convention.

**Comment by sharwell** [comment]
💭 I'm not sure I like the previous comment's behavior. It appears to preclude the use of CamelCase typing for selecting a suggested name. This may require [overlapping] be the longest sequence for which a pattern matches...
